include/vcpkg/vcpkgcmdarguments.h

```diff
@@ -29,48 +29,101 @@ namespace vcpkg
         std::vector<std::string> command_arguments;
     };
 
+    struct MetadataMessage
+    {
+        constexpr MetadataMessage() noexcept : kind(MetadataMessageKind::Unused), unused{} { }
+        /*implicit*/ constexpr MetadataMessage(const msg::MessageT<>& message) noexcept
+            : kind(MetadataMessageKind::Message), message(&message)
+        {
+        }
+
+        template<int N>
+        /*implicit*/ constexpr MetadataMessage(const char (&literal)[N]) noexcept
+            : kind(MetadataMessageKind::Literal), literal(literal)
+        {
+        }
+
+        template<class Callback,
+                 std::enable_if_t<std::is_convertible_v<const Callback&, LocalizedString (*)()>, int> = 0>
+        /*implicit*/ constexpr MetadataMessage(const Callback& callback) noexcept
+            : kind(MetadataMessageKind::Callback), callback(callback)
+        {
+        }
+
+        constexpr MetadataMessage(std::nullptr_t) = delete;
+
+        MetadataMessage(const MetadataMessage&) = delete;
+        MetadataMessage& operator=(const MetadataMessage&) = delete;
+
+        LocalizedString format() const;
+        void append_to(LocalizedString& target) const;
+        explicit operator bool() const noexcept;
+
+    private:
+        enum class MetadataMessageKind
+        {
+            Unused,
+            Message,
+            Literal,
+            Callback
+        };
+
+        MetadataMessageKind kind;
+        union
+        {
+            char unused;
+            const msg::MessageT<>* message;
```

Comment:
`MessageT<>` is a size_t, so this should probably be by value.

Reply:
If I make it by value, then none of the users of this can be constexpr because the macros that make `msgFoo` make them `extern const`:

```
C:\Dev\vcpkg-tool\src\vcpkg\commands.ci.cpp(94): error C2131: expression did not evaluate to a constant
  C:\Dev\vcpkg-tool\src\vcpkg\commands.ci.cpp(95): note: failure was caused by non-constant arguments or reference to a non-constant symbol
  C:\Dev\vcpkg-tool\src\vcpkg\commands.ci.cpp(95): note: see usage of 'vcpkg::msgCISettingsOptExclude'
C:\Dev\vcpkg-tool\src\vcpkg\commands.ci.cpp(104): error C2131: expression did not evaluate to a constant
  C:\Dev\vcpkg-tool\src\vcpkg\commands.ci.cpp(105): note: failure was caused by non-constant arguments or reference to a non-constant symbol
  C:\Dev\vcpkg-tool\src\vcpkg\commands.ci.cpp(105): note: see usage of 'vcpkg::msgCISwitchOptDryRun'
C:\Dev\vcpkg-tool\src\vcpkg\commands.ci.cpp(319): error C2131: expression did not evaluate to a constant
  C:\Dev\vcpkg-tool\src\vcpkg\commands.ci.cpp(321): note: failure was caused by non-constant arguments or reference to a non-constant symbol
  C:\Dev\vcpkg-tool\src\vcpkg\commands.ci.cpp(321): note: see usage of 'vcpkg::msgCmdCiSynopsis'
  [44/163] Building CXX object CMakeFiles\vcpkglib.dir\src\vcpkg\commands.install.cpp.obj
```

Reply:
https://github.com/microsoft/vcpkg-tool/blob/c1854e863050964c528d965b06134348a39d346d/include/vcpkg/base/messages.h#L248-L252